src/pubsub/index.js

```diff
@@ -0,0 +1,663 @@
+    const id = peerId.toB58String()
+    const existing = this.peers.get(id)
+
+    // If peer streams already exists, do nothing
+    if (existing) {
+      return existing
+    }
+
+    // else create a new peer streams
+    this.log('new peer', id)
+
+    const peerStreams = new PeerStreams({
+      id: peerId,
+      protocol
+    })
+
+    this.peers.set(id, peerStreams)
+    peerStreams.once('close', () => this._removePeer(peerId))
+
+    return peerStreams
+  }
+
+  /**
+   * Notifies the router that a peer has been disconnected.
+   * @private
+   * @param {PeerId} peerId
+   * @returns {PeerStreams | undefined}
+   */
+  _removePeer (peerId) {
+    if (!peerId) return
+    const id = peerId.toB58String()
+    const peerStreams = this.peers.get(id)
+    if (!peerStreams) return
+
+    // close peer streams
+    peerStreams.removeAllListeners()
+    peerStreams.close()
+
+    // delete peer streams
+    this.log('delete peer', id)
+    this.peers.delete(id)
+
+    // remove peer from topics map
+    for (const peers of this.topics.values()) {
+      peers.delete(id)
+    }
+
+    return peerStreams
+  }
+
+  // MESSAGE METHODS
+
+  /**
+   * Responsible for processing each RPC message received by other peers.
+   * @param {string} idB58Str peer id string in base58
+   * @param {DuplexIterableStream} stream inbound stream
+   * @param {PeerStreams} peerStreams PubSub peer
+   * @returns {Promise<void>}
+   */
+  async _processMessages (idB58Str, stream, peerStreams) {
+    try {
+      await pipe(
+        stream,
+        async (source) => {
+          for await (const data of source) {
+            const rpcBytes = data instanceof Uint8Array ? data : data.slice()
+            const rpcMsg = this._decodeRpc(rpcBytes)
+
+            this._processRpc(idB58Str, peerStreams, rpcMsg)
+          }
+        }
+      )
+    } catch (err) {
+      this._onPeerDisconnected(peerStreams.id, err)
+    }
+  }
+
+  /**
+   * Handles an rpc request from a peer
+   * @param {String} idB58Str
+   * @param {PeerStreams} peerStreams
+   * @param {RPC} rpc
+   * @returns {boolean}
+   */
+  _processRpc (idB58Str, peerStreams, rpc) {
+    this.log('rpc from', idB58Str)
+    const subs = rpc.subscriptions
+    const msgs = rpc.msgs
+
+    if (subs.length) {
+      // update peer subscriptions
+      subs.forEach((subOpt) => this._processRpcSubOpt(idB58Str, subOpt))
+      this.emit('pubsub:subscription-change', peerStreams.id, subs)
+    }
+
+    if (!this._acceptFrom(idB58Str)) {
+      this.log('received message from unacceptable peer %s', idB58Str)
+      return false
+    }
+
+    if (msgs.length) {
+      msgs.forEach(message => {
+        if (!(this.canRelayMessage || message.topicIDs.some((topic) => this.subscriptions.has(topic)))) {
+          this.log('received message we didn\'t subscribe to. Dropping.')
+          return
+        }
+        const msg = utils.normalizeInRpcMessage(message, PeerId.createFromB58String(idB58Str))
```

Comment:
The PeerId passed to `normalizeInRpcMessage` is converted to its string form for use, and I only see this called with a PeerId here. Perhaps this should just change to taking and using the string to avoid the double conversion.